spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java

```diff
@@ -93,33 +98,56 @@ public TwoStepsConversions(CustomConversions customConversions, ObjectToKeyFacto
 
 		this.internalConversionService.addConverter(BYTE_ARRAY_TO_BLOB_CONVERTER);
 		this.internalConversionService.addConverter(BLOB_TO_BYTE_ARRAY_CONVERTER);
-
 	}
 
 	@Override
 	public <T> T convertOnRead(Object val, Class targetCollectionType,
 			Class targetComponentType) {
-		return convertOnRead(val, false, targetCollectionType, targetComponentType);
+		return (T) convertOnRead(val, EmbeddedType.NOT_EMBEDDED,
+				targetCollectionType,
+				ClassTypeInformation.from(targetComponentType));
 	}
 
 	@Override
-	public <T> T convertOnReadEmbedded(Object val, Class targetCollectionType,
-			Class targetComponentType) {
-		return convertOnRead(val, true, targetCollectionType, targetComponentType);
+	public <T> T convertOnRead(Object val, EmbeddedType embeddedType,
+			TypeInformation targetTypeInformation) {
+		TypeInformation componentTypeInformation;
+		if (targetTypeInformation.isCollectionLike()) {
+			componentTypeInformation = targetTypeInformation.getComponentType();
+		}
+		else {
+			componentTypeInformation = targetTypeInformation;
+		}
+		return convertOnRead(val, embeddedType,
+				targetTypeInformation.isCollectionLike() ? targetTypeInformation.getType()
+						: null,
+				componentTypeInformation);
 	}
 
-	@SuppressWarnings("unchecked")
-	private <T> T convertOnRead(Object val, boolean isEmbedded,
-			Class targetCollectionType, Class targetComponentType) {
+	private <T> T convertOnRead(Object val, EmbeddedType embeddedType,
+			Class targetCollectionType, TypeInformation targetComponentType) {
 		if (val == null) {
 			return null;
 		}
-		BiFunction<Object, Class<?>, T> readConverter = isEmbedded
-				? this::convertOnReadSingleEmbedded
-				: this::convertOnReadSingle;
+		BiFunction<Object, TypeInformation<?>, ?> readConverter;
+		switch (embeddedType) {
+		case EMBEDDED_MAP:
+			readConverter = (x, y) -> convertOnReadSingleEmbeddedMap(x,
```

Comment:
There is no meaning to having names here. In fact the other methods in this switch-case don't even have named params at all. Some real functional languages actually just allow you to use special no-name-necessary param names for these params in situations just like this.

Reply:
The other methods have parameter names in the definition. I don't argue that sometimes it makes sense to skip the names, but in this particular case it would be more readable.